src/tsc_now.rs

```diff
@@ -92,10 +93,51 @@ impl TSCLevel {
 /// rely on the result to say tsc is stable so that no need to
 /// sync TSCs by ourselves.
 fn is_tsc_stable() -> bool {
-    let clock_source =
-        read_to_string("/sys/devices/system/clocksource/clocksource0/available_clocksource");
+    has_invariant_tsc() || clock_source_has_tsc()
+}
+
+fn clock_source_has_tsc() -> bool {
+    #[cfg(target_os = "linux")]
+    {
+        let clock_sources = [
+            "/sys/devices/system/clocksource/clocksource0/current_clocksource",
+            "/sys/devices/system/clocksource/clocksource0/available_clocksource",
+        ];
+
+        for path in clock_sources {
+            match read_to_string(path) {
+                Ok(content) => {
+                    if content.contains("tsc") {
+                        return true;
+                    }
+                }
+                Err(e) if e.kind() == ErrorKind::NotFound => continue,
+                Err(_) => return false,
```

Comment:
Perhaps tolerate all errors (just `continue`)?

Reply:
other errors should return false i think.

Reply:
Why? You test either file exists and if any one of them exist and contains "TSC", it means "TSC" is valid.

In other words, what kind of other errors are possible and how they indicate an early false?

Reply:
The current implementation logic is the same as before(`clock_source.map(|s| s.contains("tsc")).unwrap_or(false)`). Other errors, such as `PermissionDenied` should return false.

Reply:
But perviously we check only one file, and now we check two. The previous logic is the same as:

```rust
        let clock_sources = [
            "/sys/devices/system/clocksource/clocksource0/available_clocksource",
        ];

        for path in clock_sources {
            match read_to_string(path) {
                Ok(content) => {
                    if content.contains("tsc") {
                        return true;
                    }
                }
                Err(e) => continue,
            }
        }

        false
```

Reply:
The previous implementation was not entirely correct, we should check `current_clocksource` first, if `current_clocksource` is not provided, we fallback check `available`.

`current_clocksource` is the clocksource system specified by the sysadmin, if they won't let user use `tsc`, they may set `current_clocksource` to such as `hpet`.

Reply:
This is my expected code ...

```rust
        let clock_sources = [
            "/sys/devices/system/clocksource/clocksource0/current_clocksource",
            "/sys/devices/system/clocksource/clocksource0/available_clocksource",
        ];

        for path in clock_sources {
            match read_to_string(path) {
                Ok(content) => {
                    if content.contains("tsc") {
                        return true;
                    }
                }
                Err(e) => continue,
            }
        }

        false
```

The difference is that, no matter `current_clocksource` is missing, cannot access, and other errors, we can try `available_clocksource`.

I don't find the reason to refuse test `available_clocksource` when read `current_clocksource` returns some errors.

Reply:
I don't think so.

Only read `available_clocksource ` if `current_clocksource` doesn't exist, and follow `current_clocksource` as long as `current_clocksource` exists, regardless of what `available_clocksource` provides.

Imagine a scenario where the administrator has set `hpet` in `current_clocksource`, but `tsc` is provided in `available_clocksource `. Obviously, the administrator does not want to use `tsc`, and if we reads `available_clocksource `, it will be difficult to troubleshoot the problem if needed because it does not meet the administrator's expectations.

Reply:
You current code doesn't work in the way you describe. It will read `available_clocksource` if `current_clocksource` exists but doesn't contain `tsc`.

To implement the manner your describe, the code is like:

```rust
{
    if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/current_clocksource") {
        return content.contains("tsc")
    }

    if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/available_clocksource") {
        return content.contains("tsc")
    }

    false
}
```

Reply:
`available_clocksource ` is just a template file, we fallback to this iff `current_clocksource` not exists(rarely)

Reply:
> You current code doesn't work in the way you describe. It will read `available_clocksource` if `current_clocksource` exists but doesn't contain `tsc`.
> 
> To implement the manner your describe, the code is like:
> 
> ```rust
> {
>     if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/current_clocksource") {
>         return content.contains("tsc")
>     }
> 
>     if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/available_clocksource") {
>         return content.contains("tsc")
>     }
> 
>     false
> }
> ```

Correct! let me fix it.

Reply:
Or if you consider that a `current_clocksource` that permission deny should indicate a conservative resolution to return false (reject to use TSC), write:

```rust
{
    match read_to_string("/sys/devices/system/clocksource/clocksource0/current_clocksource") {
        Ok(content) => return content.contains("tsc")
        Err(e) if e.kind() == ErrorKind::NotFound => {}
        Err(_) => return false,
    }

    if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/available_clocksource") {
        return content.contains("tsc")
    }

    false
}
```

Reply:
Fixed, thx~